protomap/proto.go

```diff
@@ -18,63 +18,361 @@
+//
+// List fields are 'repeated' in the input protobuf. We have two cases of such
+// fields:
+//  1. leaf-list fields which have scalar values - and hence are mapped in the
+//     same way as the handling of individual fields in the protobuf.
+//  2. list types in YANG - we only support keyed lists, since these have their
+//     own valid paths. For the generated protobufs we create a new XXXKey message
+//     which is the repeated type. Scalar fields within that message are the
+//     individual keys of the list (there are >= 1 of them) -- and the single
+//     message type that is included is a child message.
+func parseList(fd protoreflect.FieldDescriptor, v protoreflect.Value, vals map[*gpb.Path]interface{}, basePath *gpb.Path, mapPath []*gpb.Path) error {
+	// Lists cannot map to >1 different schema path in the compressed form of generated
+	// protobufs.
+	if len(mapPath) > 1 {
+		return fmt.Errorf("invalid list, maps to >1 schema path, field: %s", fd.FullName())
+	}
+	listPath := mapPath[0]
+
+	// TODO(robjs): This handles the case where we have a list - but not a leaf-list.
+	//              Add implementation to handle leaf lists.
+	l := v.List()
+	var listVal proto.Message
+	for i := 0; i < l.Len(); i++ {
+		lv := l.Get(i)
+
+		if fd.Kind() != protoreflect.MessageKind {
+			return fmt.Errorf("invalid list, value is not a proto message, %s - is %T", fd.FullName(), lv.Interface())
+		}
+		listMsg := lv.Message().Interface().(proto.Message)
+
+		var listParseErr error
+		listKeys := map[string]string{}
+		mappedValues := map[*gpb.Path]interface{}{}
+
+		unpopRange{listMsg.ProtoReflect()}.Range(func(fd protoreflect.FieldDescriptor, v protoreflect.Value) bool {
+			pl, err := parseListField(fd, v, listPath)
+			if err != nil {
+				listParseErr = err
+				return false
+			}
+
+			// If this field was not the value of the list then we receive a populated
+			// single key back from the parseListField call. We don't check for nil
+			// since if it is not populated then we'll simply never do this mapping.
+			for k, v := range pl.keys {
+				listKeys[k] = v
+			}
+
+			// When there are keys in the list, we'll also have fields that they map to
+			// in the output set of paths, so add these to the values that we're receiving.
+			for k, v := range pl.mappedValues {
+				mappedValues[k] = v
+			}
+
+			// If this was the list value, then we return the value of the list,
+			// which is always a protobuf message back.
+			if pl.member != nil {
+				listVal = pl.member
+			}
+			return true
+		})
+
+		if listParseErr != nil {
+			return fmt.Errorf("could not parse a field within the list %s , %v", fd.FullName(), listParseErr)
+		}
+
+		// Handle recursing down the tree for the list that we just extracted.
+		p := proto.Clone(listPath).(*gpb.Path)
+		if basePath != nil {
+			// This is the first time that we have found a path that requires a
+			// data tree path, not a schema tree path.
+			p = resolvedPath(basePath, listPath)
+		}
```

Comment:
I think having the `if basePath != nil` check within `resolvedPath` makes it slightly more convenient as I see this being written in another place as well.

If clone is also removed, that line and this block can just become `p := resolvedPath(basePath, listPath)`.